dev/src/bulk-writer.ts

```diff
@@ -326,34 +281,99 @@ class BulkCommitBatch {
   }
 }
 
+/**
+ * The error thrown when a BulkWriter operation fails.
+ *
+ * @class BulkWriterError
+ */
+export class BulkWriterError extends Error {
+  /** @hideconstructor */
+  constructor(
+    /** The status code of the error. */
+    readonly code: GrpcStatus,
+
+    /** The error message of the error. */
+    readonly message: string,
+
+    /** The document reference the operation was performed on. */
+    readonly documentRef: firestore.DocumentReference,
+
+    /** The type of operation performed. */
+    readonly operationType: 'create' | 'set' | 'update' | 'delete',
+
+    /** How many times this operation has been retried. */
+    readonly retryCount: number
+  ) {
+    super(message);
+  }
+}
+
 /**
  * A Firestore BulkWriter than can be used to perform a large number of writes
  * in parallel. Writes to the same document will be executed sequentially.
  *
- * @class
+ * @class BulkWriter
  */
 export class BulkWriter {
   /**
    * The maximum number of writes that can be in a single batch.
    */
-  private maxBatchSize = MAX_BATCH_SIZE;
+  private _maxBatchSize = MAX_BATCH_SIZE;
 
   /**
    * A queue of batches to be written.
    */
-  private batchQueue: BulkCommitBatch[] = [];
+  private _batchQueue: BulkCommitBatch[] = [];
+
+  private _retryBatchQueue: BulkCommitBatch[] = [];
+
+  /**
+   * A list of promises that represent sent batches. Each promise is resolved
+   * when the batch's response is received.
+   */
+  private _pendingBatches: Set<Promise<void>> = new Set();
```

Comment:
If you get creative, we should indicate via the naming that `_pendingBatches` contains both `_batchQueue` and `_retryBatchQueue`.

Reply:
Couldn't think of a good way for naming, but added more detailed comments.